lib/ansible/modules/network/fortimanager/fmgr_fwobj_address.py

```diff
@@ -775,84 +599,64 @@ def main():
-        response = fmg.login()
-        if response[1]['status']['code'] != 0:
-            module.fail_json(msg="Connection to FortiManager Failed")
-    except Exception:
-        module.fail_json(msg="Connection to FortiManager Failed")
+    module.paramgram = paramgram
+    fmgr = None
+    if module._socket_path:
+        connection = Connection(module._socket_path)
+        fmgr = FortiManagerHandler(connection, module)
+        fmgr._tools = FMGRCommon()
     else:
-        # START SESSION LOGIC
-        # MODULE PARAMGRAM
-        paramgram = {
-            "adom": module.params["adom"],
-            "allow-routing": module.params["allow_routing"],
-            "associated-interface": module.params["associated_interface"],
-            "cache-ttl": module.params["cache_ttl"],
-            "color": module.params["color"],
-            "comment": module.params["comment"],
-            "country": module.params["country"],
-            "end-ip": module.params["end_ip"],
-            "fqdn": module.params["fqdn"],
-            "name": module.params["name"],
-            "start-ip": module.params["start_ip"],
-            "visibility": module.params["visibility"],
-            "wildcard": module.params["wildcard"],
-            "wildcard-fqdn": module.params["wildcard_fqdn"],
-            "ipv6": module.params["ipv6"],
-            "ipv4": module.params["ipv4"],
-            "group_members": module.params["group_members"],
-            "group_name": module.params["group_name"],
-            "ipv4addr": module.params["ipv4addr"],
-            "ipv6addr": module.params["ipv6addr"],
-            "multicast": module.params["multicast"],
-            "mode": module.params["mode"],
-            "obj-id": module.params["obj_id"],
-        }
-
-        if paramgram["adom"] is None:
-            paramgram["adom"] = "root"
-        if paramgram["mode"] is None:
-            paramgram["mode"] = "add"
-        if paramgram["color"] is None:
-            paramgram["color"] = 22
-        if paramgram["comment"] is None:
-            paramgram["comment"] = "Created by Ansible"
-        if paramgram["allow-routing"] is None:
-            paramgram["allow-routing"] = "disable"
-        if paramgram["visibility"] is None:
-            paramgram["visibility"] = "enable"
+        module.fail_json(**FAIL_SOCKET_MSG)
 
+    results = DEFAULT_RESULT_OBJ
+    try:
         if paramgram["ipv4"] is not None and paramgram["ipv6"] is None and paramgram["multicast"] is None:
             # PROCESS IPv4
-            results = fmgr_fwobj_ipv4(fmg, paramgram)
-            fmgr_logout(fmg, module, results=results, good_codes=[0, -2, -3])
+            results = fmgr_fwobj_ipv4(fmgr, paramgram)
+            fmgr.govern_response(module=module, results=results,
+                                 ansible_facts=fmgr.construct_ansible_facts(results, module.params, paramgram))
 
         if paramgram["ipv4"] is None and paramgram["ipv6"] is not None and paramgram["multicast"] is None:
```

Comment:
Minor suggestion - seems like both line 646 and 652 are checking for `paramgram["ipv4"]` to be `None`, can they be clubbed into a single check and then based on the values of paramgram['ipv6'] and paramgram['multicast'] the correct logic can be determined?

Reply:
Thanks for the repetitive catch. The original idea was to make sure that if you were using one, the other two were always empty... validation. Instead moved this logic to mutually_exclusive=[] and rearranged the specified block.